evm/src/invariant_fuzz.rs

```diff
@@ -0,0 +1,363 @@
+//! Fuzzing support abstracted over the [`Evm`](crate::Evm) used
+use crate::fuzz::*;
+
+use crate::fuzz::strategies::fuzz_param;
+// use core::slice::SlicePattern;
+use ethers::{
+    abi::{Abi, Function},
+    types::{Address, Bytes, U256},
+};
+pub use proptest::test_runner::Config as FuzzConfig;
+use proptest::{
+    prelude::*,
+    test_runner::{TestError, TestRunner},
+};
+use revm::db::DatabaseRef;
+use serde::{Deserialize, Serialize};
+use std::{cell::RefCell, collections::BTreeMap};
+
+use crate::executor::{Executor, RawCallResult};
+
+/// Wrapper around any [`Evm`](crate::Evm) implementor which provides fuzzing support using [`proptest`](https://docs.rs/proptest/1.0.0/proptest/).
+///
+/// After instantiation, calling `fuzz` will proceed to hammer the deployed smart contract with
+/// inputs, until it finds a counterexample. The provided `TestRunner` contains all the
+/// configuration which can be overridden via [environment variables](https://docs.rs/proptest/1.0.0/proptest/test_runner/struct.Config.html)
+pub struct InvariantExecutor<'a, DB: DatabaseRef + Clone> {
+    // evm: RefCell<&'a mut E>,
+    /// The VM todo executor
+    pub evm: &'a mut Executor<DB>,
+    runner: TestRunner,
+    sender: Address,
+    contracts: &'a BTreeMap<Address, (String, Abi)>,
+}
+
+impl<'a, DB> InvariantExecutor<'a, DB>
+where
+    DB: DatabaseRef + Clone,
+{
+    /// Instantiates a fuzzed executor EVM given a testrunner
+    pub fn new(
+        evm: &'a mut Executor<DB>,
+        runner: TestRunner,
+        sender: Address,
+        contracts: &'a BTreeMap<Address, (String, Abi)>,
+    ) -> Self {
+        Self { evm, runner, sender, contracts }
+    }
+
+    /// Fuzzes the provided function, assuming it is available at the contract at `address`
+    /// If `should_fail` is set to `true`, then it will stop only when there's a success
```

Comment:
This sounds like it refers to something like `testFail`, but we probably don't need to support `invariantFail` style tests—is there a use case for that?